### ISSUE TITLE
ViewModelScoped Dependency as a constructor argument

### ISSUE THREAD

**Issue by mahmed1987** [InsertKoinIO/koin#1871]
I want to introduce a dependency in my ViewModel that is tied to the ViewModel (its life and death should work with the life and death of the ViewModel) 

This code (copied from the official koin resource) works

```
class MyScopeViewModel : ViewModel(), KoinScopeComponent {

    override val scope: Scope = createScope(this)

    // inject your dependency
    val session by scope.inject<Session>()

    // clear scope
    override fun onCleared() {
        super.onCleared()
        scope.close()
    }
}
```
However this is not what I want to achieve , I want to inject the Session class as a Constructor argument like this

```
class MyScopeViewModel(val session:Session) : ViewModel(){





}
```
The above responds with the exception stating that the Session bean could not be found.

What am i missing here ?

Koin Version - 3.5.0

**Comment by arnaudgiuliani** [comment]
For now we can handle scope creation by field: https://insert-koin.io/docs/reference/koin-android/scope#viewmodel-scope-since-354

Another solution is in study for 3.6

**Comment by mahmed1987** [comment]
yeah @arnaudgiuliani  , this appears to be quite necessary to have dependencies injected that are scoped to the ViewModel itself  , and not single<> (kinda like what ScopedViewModel does , but yeah would want to do that in a normal androidx.ViewModel) 

A use case can be for example a facility to sort a student list in a StudentViewModel called StudentSorter(). The StudentSorter class doesn't need to survive StudentViewModel death. 

Great to know that this is in the radar .

**Comment by uragiristereo** [comment]
You can have another ViewModel that specifically provides only the scope, kinda hacky but works.

Suppose that you have this ViewModel:
```kt
class FeatureFooViewModel(val session: Session) : ViewModel() {

}
```

1. Define the ViewModel, you should only define it once.
```kt
class ScopeProvider : ViewModel, KoinScopeComponent {
    override val scope = createScope(this)

    override fun onCleared() {
        super.onCleared()
        scope.close()
    }
}
```

2. Scope the required dependencies to the ViewModel.
```kt
val featureFooModule = module {
    scope<ScopeProvider> {
        scopedOf(::Session)
        viewModelOf(::FeatureFooViewModel)
    }
}
```

3. Implement a function to create the scope.
```kt
fun Fragment.viewModelScope(ownerProducer: () -> ViewModelStoreOwner = { this }): Lazy<Scope> {
    return lazy {
        // notice that I'm using the vanilla ViewModel creator here
        val scopeProvider by viewModels<ScopeProvider>(ownerProducer)

        scopeProvider.scope
    }
}
```

4. Finally, consume the scope.
```kt
class FooFragment(), AndroidScopeComponent {
    override val scope by viewModelScope()

    private val fooViewModel by viewModel<FooViewModel>()
}
```

The best part is that you can scope the dependencies by the navigation graph parent instead (for androidx-navigation) by passing the parent `NavBackStackEntry` into the `ownerProducer` lambda.

**Comment by est7** [comment]
Does this help you? https://github.com/InsertKoinIO/koin/issues/1821

**Comment by mahmed1987** [comment]
hey @est7  , it does not actually . The issue discussed there is different then what i was looking for . @arnaudgiuliani  suggests that a solution is in the study in version 3.6

**Comment by arnaudgiuliani** [comment]
ViewModel scope + constructor injection will be possible: https://github.com/InsertKoinIO/koin/pull/2181
Also make `viewModelScope()` function lighter + scope closure possible

stay tunned for next release 👍

**Comment by psucjuerkevin** [comment]
#

> ViewModel scope + constructor injection will be possible: [#2181](https://github.com/InsertKoinIO/koin/pull/2181) Also make `viewModelScope()` function lighter + scope closure possible
> 
> stay tunned for next release 👍

Hi @arnaudgiuliani 
First of all, thanks for fixing this issue, that's what I've been waiting for.  I appreciate it.
Secondly, could the annotation also achieve the same purpose?
As I've known, these isn't annotation that can generate  `viewModelScope ` rather `scope<MyScopeViewModel>`

Manual way:
` viewModelScope {`
      `    viewModelOf(::MyScopeViewModel)`
      `    scopedOf(::Session)`
      `    scopedOf(::SessionConsumer)`
`}`

Annotation way (generated automatically):
` scope<MyScopeViewModel> {`
      `    scopedOf(::Session)`
      `    scopedOf(::SessionConsumer)`
`}`
